### PR TITLE
test: add integration test for accepting an invite to an empty room

### LINKED ISSUE #4642: Impossible to decline an invite to an empty room
- Create a private room
- invite someone
- make them admin too (not sure if this is necessary)
- leave the room before the other person can accept the invite

Result:
- The other user now has an invite in the room list, but it can't be accepted or declined and never goes away `403 FORBIDDEN`

### PR DESCRIPTION
**Pull request by vektrace**
This PR is related to #4642. I have added the second of the two requested integration tests, according to which accepting an invite did indeed not work, but instead threw a completely different error than mentioned.

- [ ] I've documented the public API changes in the appropriate changelog files (see [Writing changelog entries][pull-request-template-changelog]).
- [ ] This PR was made with the help of AI.

[pull-request-template-changelog]: https://github.com/matrix-org/matrix-rust-sdk/blob/main/CONTRIBUTING.md#writing-changelog-entries

Signed-off-by: Paul8711 <154304377+paul8711-code@users.noreply.github.com>

### CODE REVIEW

**Comment by vektrace** [comment]
@Hywan could you also take a look at this PR? It too is relevant to our discussion in the mentioned issue.

File: testing/matrix-sdk-integration-testing/src/tests/room.rs [hunk 1]
```diff
@@ -36,6 +36,52 @@ use tracing::{debug, error, warn};
...
+        }))
+        .await?
+        .room_id()
+        .to_owned();
+
+    if let Some(room) = alice.get_room(&room_id) {
+        room.leave().await?;
+    }
+
+    if let Some(room) = bob.get_room(&room_id) {
```
Comment (Hywan) [inline testing/matrix-sdk-integration-testing/src/tests/room.rs]: At this step, you're not sure Bob has received the invite. You must ensure the sync has received the invite before trying to get the room.

**Review by Hywan — COMMENTED** [review]
Thanks for the test! I think we need to improve it a bit though.

File: testing/matrix-sdk-integration-testing/src/tests/room.rs
Comment (Hywan) [inline testing/matrix-sdk-integration-testing/src/tests/room.rs]: Here too, I think we must ensure the room exists to ensure it's left.

```rust
let room = alice.get_room(&room_id).expect("Alice must see the room");
room.leave().await?;
```

**Review by Hywan — DISMISSED** [review]

**Comment by Hywan** [comment]
The new test was failing. It is a bit annoying. Let's see if re-running the CI will fix it.

**Comment by vektrace** [comment]
If you need/want me to do anything else, just let me know. I will only be partially available in the coming week, though.

**Comment by Hywan** [comment]
CI is still failing. Can you fix the test please? Do you need help?

**Comment by vektrace** [comment]
Well, I do have one question: What would be the expected behaviour when attempting to accept an invite to an empty room? Should the invite be discarded?

**Comment by Hywan** [comment]
Well, good question!

First off, this is a server issue. Someone creates a room, invite you, then leave the room: you cannot join the room because it's empty. It's a long-standing bug. See for example, https://github.com/element-hq/synapse/issues/19034.

Second, from a client point of view, I suspect we will get an error from the server, so we can have a rule like:

- if we accept the invite
- but we get a specific error from the server ([such as a 403](https://spec.matrix.org/v1.19/client-server-api/#post_matrixclientv3roomsroomidjoin), maybe with an `M_FORBIDEN`?)
- we can remove the room from the list of rooms
  - to do that, we can try to forget the room (with `Client::forget_room`), I don't know if that would work, [the spec about `/forget`](https://spec.matrix.org/v1.19/client-server-api/#post_matrixclientv3roomsroomidforget) doesn't seem to say it wouldn't work, so I think it's our best solution!

**Comment by vektrace** [comment]
Thank you very much! I will try that as soon as I have time.

**Comment by vektrace** [comment]
I _think_ I understand why the response is `M_UNKNOWN` instead of `M_FORBIDDEN`. I checked almost everything (only partially the Synapse source code). The homeserver the bug was reported with **could** be newer (so would be the error messages) than our version of Synapse for tests. This is problematic because if I add the test checking for the `403 M_FORBIDDEN` error, the checks will still fail.
I also heard that Synapse's invites are currently a bit broken, so that could be it too.

_Note that this is just my theory and could be completely wrong._

> EDIT: This theory was, in fact, completely wrong.

**Comment by Hywan** [comment]
If you need to update Synapse, feel free to do so:

https://github.com/matrix-org/matrix-rust-sdk/blob/b641f505f7ed5af163c6948897ad714c222660e9/testing/matrix-sdk-integration-testing/assets/Dockerfile#L1

**Comment by vektrace** [comment]
Accepting the invite returns `404 M_UNKNOWN` with "Can't join remote room because no servers that are in the room have been provided." rather than `403 M_FORBIDDEN`. Should we call `forget_room` on 404, or are we targeting only `403 M_FORBIDDEN`?

**Comment by Hywan** [comment]
I think `M_UNKNOWN` is fine.

File: testing/matrix-sdk-integration-testing/src/tests/room.rs [hunk 2]
```diff
@@ -79,7 +80,21 @@ async fn test_empty_room_accept_invite() -> Result<()> {
         if let Some(room) = bob.get_room(&room_id)
             && matches!(room.state(), RoomState::Invited)
         {
-            room.join().await?;
+            if let Err(err) = room.join().await {
+                if let Some(api_err) = err.as_client_api_error() {
+                    if api_err.status_code == StatusCode::NOT_FOUND {
+                        // leave first
+                        room.leave().await?;
```
Comment (vektrace) [inline testing/matrix-sdk-integration-testing/src/tests/room.rs]: Leaving is necessary as forgetting a room only works if we were banned or left the room. I could also remove `room.forget()` if needed/wanted.

File: testing/matrix-sdk-integration-testing/src/tests/room.rs [hunk 3]
```diff
@@ -90,6 +91,74 @@ async fn test_empty_room_decline_invite() -> Result<()> {
...
+            && matches!(room.state(), RoomState::Invited)
+        {
+            if let Err(err) = room.join().await {
+                if let Some(api_err) = err.as_client_api_error() {
+                    if api_err.status_code == StatusCode::NOT_FOUND {
+                        // leave first
+                        room.leave().await?;
+
+                        // then forget
+                        room.forget().await?;
```
Comment (poljar) [inline testing/matrix-sdk-integration-testing/src/tests/room.rs]: Can't we return an Ok here directly instead of going the roundabout way with this?

The place where the `ensure!()` call is would just return an error in that case.

The boolean then becomes unnecessary.

**Review by poljar — CHANGES_REQUESTED** [review]
Sorry, had to take over as the previous reviewer is on holiday.

File: testing/matrix-sdk-integration-testing/src/tests/room.rs (hunk 3 above)
Comment (vektrace) [reply testing/matrix-sdk-integration-testing/src/tests/room.rs]: You are absolutely right, thanks for pointing this out. I have implemented that in my latest commit.

**Review by poljar — APPROVED** [review]

**Comment by vektrace** [comment]
Thank you both for taking your time to review this!

**Comment by Hywan** [comment]
I think now we need to update the SDK to fix this problem client-side while waiting on the Matrix spec to improve. Thoughts @manuroe?
